pyop2/types/set.py

```diff
@@ -345,10 +333,17 @@ def _argtypes_(self):
     def _wrapper_cache_key_(self):
         return self.parent._wrapper_cache_key_ + (self.constant_layers, )
 
-    def __getattr__(self, name):
-        """Returns a :class:`Set` specific attribute."""
-        value = getattr(self._parent, name)
-        return value
+    @utils.cached_property
```

Comment:
I don't think that we need `cached_property` for pass-through properties like this. If this is an expensive thing to compute then we should cache it for `self._parent.sizes`, not here.

Reply:
Here I am not sure. I couldn't find a way to make it work with just `@property`. The error message looks like this:
```bash
__________________________________________________________ TestSubsetAPI.test_empty_subset ___________________________________________________________

self = <test_api.TestSubsetAPI object at 0x7fc13b2facb0>, set = Set((5, 5, 5), 'foo')

    def test_empty_subset(self, set):
        "Subsets can be empty."
>       ss = op2.Subset(set, [])

test/unit/test_api.py:328: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
<decorator-gen-21>:2: in __init__
    ???
pyop2/utils.py:133: in wrapper
    return f(*args, **kwargs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = Subset(Set((5, 5, 5), 'foo'), array([], dtype=int32)), superset = Set((5, 5, 5), 'foo'), indices = array([], dtype=float64)

    @utils.validate_type(('superset', Set, TypeError),
                         ('indices', (list, tuple, np.ndarray), TypeError))
    def __init__(self, superset, indices):
        self.comm = mpi.internal_comm(superset.comm)
    
        # sort and remove duplicates
        indices = np.unique(indices)
        if isinstance(superset, Subset):
            # Unroll indices to point to those in the parent
            indices = superset.indices[indices]
            superset = superset.superset
        assert type(superset) is Set or type(superset) is ExtrudedSet, \
            'Subset construction failed, should not happen'
    
        self._superset = superset
        self._indices = utils.verify_reshape(indices, dtypes.IntType, (len(indices),))
    
        if len(self._indices) > 0 and (self._indices[0] < 0 or self._indices[-1] >= self._superset.total_size):
            raise ex.SubsetIndexOutOfBounds(
                'Out of bounds indices in Subset construction: [%d, %d) not [0, %d)' %
                (self._indices[0], self._indices[-1], self._superset.total_size))
    
>       self._sizes = ((self._indices < superset.core_size).sum(),
                       (self._indices < superset.size).sum(),
                       len(self._indices))
E       AttributeError: can't set attribute '_sizes'

pyop2/types/set.py:418: AttributeError
```
Since the problem occured in utils.py I changed it to `@utils.cached_property` and it worked.

Reply:
Ah. I think we are running into one of the other big design issues with PyOP2 which is that the inheritance is super confusing. I don't think we need to be too fussy here.

Reply:
So to be clear, you can leave it like this.